xonsh/completer.py

```diff
@@ -396,30 +417,32 @@ def attr_complete(self, prefix, ctx):
         return attrs
 
     def _all_commands(self):
+        """Possibly update our set of cached commands and return the set."""
         path = builtins.__xonsh_env__.get('PATH', [])
-        # did PATH change?
+        # Did PATH change?
         path_hash = hash(tuple(path))
         cache_valid = path_hash == self._path_checksum
         self._path_checksum = path_hash
-        # did aliases change?
+        # Did aliases change?
         al_hash = hash(tuple(sorted(builtins.aliases.keys())))
         self._alias_checksum = al_hash
         cache_valid = cache_valid and al_hash == self._alias_checksum
         pm = self._path_mtime
-        # did the contents of any directory in PATH change?
-        for d in filter(os.path.isdir, path):
+        # Did the contents of any directory in PATH change?
+        for d in (p for p in path if os.path.isdir(p)):
             m = os.stat(d).st_mtime
             if m > pm:
                 pm = m
                 cache_valid = False
-        self._path_mtime = pm
-        if cache_valid:
-            return self._cmds_cache
-        allcmds = set()
-        for d in filter(os.path.isdir, path):
-            allcmds |= set(os.listdir(d))
-        allcmds |= set(builtins.aliases.keys())
-        self._cmds_cache = frozenset(allcmds)
+
+        if not cache_valid:
```

Comment:
> The original sequence of statements is suboptimal from both cognitive and performance perspectives.

I don't think that these two implementations have significantly different performance characteristics, though I could be wrong. In any event, I find this version less readable because of the removal of the filter, the increased average indentation level, and the extra vertical whitespace.